### ISSUE TITLE
I want the ability to call C functions 

### ISSUE THREAD

**Issue by KingBes** [zephir-lang/zephir#2439]
Greeting.zep

```zep
namespace Utils;

#include "func.h"

class Greeting
{

    public static function say()
    {
        C.hello();
    }

}

```

func.h

```c
void hello(){

}
```

**Comment by Jeckerson** [comment]
You can achieve this via optimizers - https://docs.zephir-lang.com/latest/optimizers/

**Comment by KingBes** [comment]
> You can achieve this via optimizers - https://docs.zephir-lang.com/latest/optimizers/

Can this method call the c library?

**Comment by Jeckerson** [comment]
@KingBes Yes, but not with the `#include "func.h" + C.hello()` shape from your snippet. There's no `C.foo()` syntax in Zephir. The optimizer path is the supported way and it can call any C code (your own .c/.h, a system library, anything that links into the extension).

You need three pieces:
1. Get the C code into the extension. Either compile your own C file alongside the extension, or link against an existing library, via config.json:

```json
{
  "extra-sources": ["utils/func.c"],
  "extra-libs": "-L/usr/local/lib -lmylib"
}
```

- extra-sources paths are relative to `ext/`. Put func.c/func.h under `ext/utils/` and they'll be compiled into the `.so`.
- extra-libs is forwarded verbatim to the linker. Use this for system libraries (e.g. `-lcurl`, `-lpq`).

2. Write a custom optimizer. An optimizer is a small PHP class that turns a Zephir-side function call into a C function-call expression. Drop it in a directory (e.g. `optimizers/FunctionCall/HelloOptimizer.php`):

```php
<?php
namespace MyExt\Optimizers\FunctionCall;

use Zephir\Call;
use Zephir\CompilationContext;
use Zephir\CompiledExpression;
use Zephir\Optimizers\OptimizerAbstract;

class HelloOptimizer extends OptimizerAbstract
{
    public function optimize(array $expression, Call $call, CompilationContext $context)
    {
        // Bring in your header so the generated .c can see hello()'s prototype.
        $context->headersManager->add('utils/func');

        // No params in your example; if there were, resolve them via
        // $call->getResolvedParams($expression['parameters'], $context, $expression)
        // and stitch their codes into the emitted call.

        return new CompiledExpression(
            'null',          // or 'int' / 'string' / 'bool' / 'variable' — your func's return type
            'hello()',       // exact C expression to emit
            $expression
        );
    }
}
```
The Zephir kernel's own built-ins are written exactly like this, see `src/Optimizers/FunctionCall/ASinOptimizer.php` (one-line, maps asin(x) to libc asin(x) and pulls in <math.h> via `headersManager->add('math')`). The same machinery works for arbitrary C symbols.

3. Register the optimizer directory
Also in config.json:

```json
{
    "optimizer-dirs": ["./optimizers"]
}
```

Zephir loads PHP classes from <dir>/FunctionCall/*.php and matches them to call sites by stripping the Optimizer suffix and lowercasing, so HelloOptimizer intercepts calls to hello(...).
Then call it from Zephir like a normal function

```zephir
namespace Utils;

class Greeting
{
    public static function say()
    {
        hello();   // ← matched by HelloOptimizer, emits a direct C call
    }
}
```

What about the #include "func.h" line in the original `.zep`?
Zephir parses #include directives and inlines them into the generated .c, but that alone doesn't give Zephir source code any way to call the symbols declared in that header — Zephir resolves function calls through optimizers (or its built-in function tables), not by scanning headers. So #include "func.h" without the optimizer leaves you with a header pulled in, but no Zephir-side call site that knows to emit hello().

TL;DR
extra-sources / extra-libs get your C code linked in. A custom optimizer under optimizer-dirs is what lets you actually call it from .zep. And yes, that fully covers third-party C libraries too.
